### PR TITLE
Custom widget interaction events

### LINKED ISSUE #2455: Sphere widget constant update
Hey,

I would have one question about changing the _sphere_widget_ observer from `EndInteractionEvent` to `InteractionEvent`. 

The reason why I am asking this is, because we are using _sphere_widgets_ in our package to move objects around the 3D display to enable the user to position them based on his requirements. We are also "snapping" objects to the mesh to help the user. But with the current  `EndInteractionEvent` the transducer only snaps when the interaction is finished. 

By changing the InteractionEvent, we could have it interactive all the time. We already tried that by hard-coding it in pyvista, as we are unable to change the state of the observer once it's initiated. 

Here how the interaction would like on your example:
feature_1

Here you can see how the interaction would look like on our application:
feature

Do you have an idea how to change that after the _sphere_widget_ is created? Or if that could be an additional option on your side? 

Best regards,
Tomaž

_Originally posted by @tomazbregar in https://github.com/pyvista/pyvista/discussions/2361_

### PR DESCRIPTION
**Pull request by banesullivan**
### Overview

Resolves #2455 by enabling users to specify custom interaction events for triggering the callbacks associated with widgets

cc @tomazbregar, please review if you have the opportunity

### Example

Using `vtk.vtkCommand.InteractionEvent`

```py
import numpy as np
from scipy.interpolate import griddata
import vtk
import pyvista as pv


def get_colors(n):
    """A helper function to get n colors"""
    from itertools import cycle

    import matplotlib

    cycler = matplotlib.rcParams['axes.prop_cycle']
    colors = cycle(cycler)
    colors = [next(colors)['color'] for i in range(n)]
    return colors


# Create a grid to interpolate to
xmin, xmax, ymin, ymax = 0, 100, 0, 100
x = np.linspace(xmin, xmax, num=25)
y = np.linspace(ymin, ymax, num=25)
xx, yy, zz = np.meshgrid(x, y, [0])

# Make sure boundary conditions exist
boundaries = np.array([[xmin, ymin, 0], [xmin, ymax, 0], [xmax, ymin, 0], [xmax, ymax, 0]])

# Create the PyVista mesh to hold this grid
surf = pv.StructuredGrid(xx, yy, zz)

# Create some initial perturbations
# - this array will be updated inplace
points = np.array([[33, 25, 45], [70, 80, 13], [51, 57, 10], [25, 69, 20]])


# Create an interpolation function to update that surface mesh
def update_surface(point, i):
    points[i] = point
    tp = np.vstack((points, boundaries))
    zz = griddata(tp[:, 0:2], tp[:, 2], (xx[:, :, 0], yy[:, :, 0]), method='cubic')
    surf.points[:, -1] = zz.ravel(order='F')
    return


# Get a list of unique colors for each widget
colors = get_colors(len(points))

# Begin the plotting routine
p = pv.Plotter(notebook=0)

# Add the surface to the scene
p.add_mesh(surf, color=True)

# Add the widgets which will update the surface
p.add_sphere_widget(update_surface, center=points, color=colors, radius=3, interaction_event=vtk.vtkCommand.InteractionEvent)
# Add axes grid
p.show_grid()

# Show it!
p.show()
```

### CODE REVIEW

**Comment by akaszynski** [comment]
Would be great if we can add this example here as well (probably a gallery example).

**Comment by banesullivan** [comment]
I added a demo of the custom interaction event to the slicing example -- that was the simplest and the concept/result is the same for all examples

File: pyvista/plotting/widgets.py [hunk 1]
```diff
@@ -1742,6 +1789,9 @@ def add_sphere_widget(
             If ``True``, run the callback function after the widget is
             created.
 
+        interaction_event : vtk.vtkCommand
```
Comment (akaszynski) [inline pyvista/plotting/widgets.py:1792]:
```suggestion
        interaction_event : vtk.vtkCommand, optional
```

File: pyvista/plotting/widgets.py [hunk 2]
```diff
@@ -95,6 +96,9 @@ def add_box_widget(
             If ``True``, the widget will be passed as the last
             argument of the callback.
 
+        interaction_event : vtk.vtkCommand
```
Comment (akaszynski) [inline pyvista/plotting/widgets.py:99]:
```suggestion
        interaction_event : vtk.vtkCommand, optional
```

File: pyvista/plotting/widgets.py [hunk 3]
```diff
@@ -345,6 +355,9 @@ def add_plane_widget(
             rotating the normal. This is forced to ``False`` when
             ``assign_to_axis`` is set.
 
+        interaction_event : vtk.vtkCommand
```
Comment (akaszynski) [inline pyvista/plotting/widgets.py:358]:
```suggestion
        interaction_event : vtk.vtkCommand, optional
```

File: pyvista/plotting/widgets.py [hunk 4]
```diff
@@ -203,6 +208,9 @@ def add_mesh_clip_box(
         crinkle : bool, optional
             Crinkle the clip by extracting the entire cells along the clip.
 
+        interaction_event : vtk.vtkCommand
```
Comment (akaszynski) [inline pyvista/plotting/widgets.py:211]:
```suggestion
        interaction_event : vtk.vtkCommand, optional
```

File: pyvista/plotting/widgets.py [hunk 5]
```diff
@@ -662,6 +681,9 @@ def add_mesh_slice(
             effectively disabled. This prevents the user from rotating the
             normal. This is forced to ``False`` when ``assign_to_axis`` is set.
 
+        interaction_event : vtk.vtkCommand
```
Comment (akaszynski) [inline pyvista/plotting/widgets.py:684]:
```suggestion
        interaction_event : vtk.vtkCommand, optional
```

**Comment by banesullivan** [comment]
In the docstrings, I annotated the type as `vtk.vtkCommand`, but they are really `vtk.vtkCommand.EventIds`

```py
>>> type(vtk.vtkCommand.EndInteractionEvent)
<class 'vtkCommonCorePython.vtkCommand.EventIds'>
```

**Review by akaszynski — DISMISSED** [review]
Still LGTM.

**Comment by banesullivan** [comment]
@akaszynski and @tkoyama010, Will this automatically merge in 24 hours if no further action is taken?

**Comment by akaszynski** [comment]
> @akaszynski and @tkoyama010, Will this automatically merge in 24 hours if no further action is taken?

I wish. We could probably create a GitHub action cron job that merges approved PRs after they've been open for 24H, but for now I've been doing it manually at the 24H+ mark.

**Comment by banesullivan** [comment]
Good enough for me - no rush here

**Review by akaszynski — APPROVED** [review]
Review removed due to merge with main, but merge was trivial.

Going to bypass 24H rule.

**Pull request merged**